kernel/src/actions/deletion_vector.rs

```diff
@@ -223,15 +229,111 @@ pub(crate) fn treemap_to_bools(treemap: RoaringTreemap) -> Vec<bool> {
     }
 }
 
+#[derive(PartialEq)]
+#[repr(C)]
+pub enum VectorType {
+    Selection,
+    Deletion,
+}
+
+pub trait BitSet<V> {
+    fn set_bit(&mut self, idx: usize, value: impl Into<V>);
+}
+
+pub trait ContainerIterator<C, V>
+where
+    C: BitSet<V>,
+{
+    fn next_batch(&mut self, mut container: C, batch_size: usize) -> Option<C> {
+        self.next_batch_with(&mut container, batch_size)
+            .map(|_| container)
+    }
+    fn next_batch_with(&mut self, container: &mut C, batch_size: usize) -> Option<()>;
+}
+
+impl<C, V> BitSet<V> for C
+where
+    C: AsMut<[V]>,
+{
+    fn set_bit(&mut self, idx: usize, value: impl Into<V>) {
+        if let Some(i) = self.as_mut().get_mut(idx) {
+            *i = value.into();
+        } else {
+            tracing::info!("Index value in DV: {} is None", idx);
+        }
+    }
+}
+
+#[repr(C)]
+pub struct DeletionVectorBatch {
+    inner_iter: Peekable<IntoIter>,
+    position: usize,
+    total_rows: usize,
+    vector_type: VectorType,
+   
+}
+
+impl DeletionVectorBatch {
+    pub fn new(treemap: RoaringTreemap, total_rows: usize, vector_type: VectorType) -> Self {
+        let inner_iter = treemap.into_iter().peekable();
+        Self {
+            inner_iter,
+            position: 0,
+            total_rows,
+            vector_type,
+        }
+    }
+
+    pub fn full_vector(&mut self) -> Option<Vec<bool>> {
+        self.next_batch(vec![false; self.total_rows], self.total_rows)
+    }
+    
+    pub fn full_bit_vector(&mut self) -> Option<Vec<u8>> {
+        self.next_batch(vec![0; self.total_rows], self.total_rows)
+    }
+
+    #[inline]
+    fn flip_value(&self) -> bool {
+        self.vector_type == VectorType::Deletion
+    }
+}
+
+impl<C, V> ContainerIterator<C, V> for DeletionVectorBatch
+where
+    C: BitSet<V> + AsMut<[V]>,
+    V: From<bool> + Copy,
+{
+    fn next_batch_with(&mut self, container: &mut C, batch_size: usize) -> Option<()> {
+        if self.position >= self.total_rows {
+            return None;
+        }
+        let slice_range = self.position as u64 + batch_size as u64;
+
+        while let Some(bit) = self.inner_iter.next() {
```

Comment:
I don't think this is correct -- if the first available index is out of range, it will get lost. We probably need something like this ([Peekable::if_next](https://doc.rust-lang.org/std/iter/struct.Peekable.html#method.next_if) FTW!):
```rust
while let Some(index) = self.inner_iter.if_next(|index| index <= batch_end) {
    container.set_bit(index as usize - self.position, self.flip_value());
}
```

Reply:
irrelevant now